crates/jstzd/src/runnable_image/mod.rs

```diff
@@ -0,0 +1,249 @@
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    net::IpAddr,
+    sync::Arc,
+};
+
+use bollard::{
+    container::{Config, CreateContainerOptions},
+    secret::{HostConfig, PortBinding, PortMap},
+    Docker,
+};
+
+use crate::{container::ContainerId, image::Image};
+
+pub mod builder;
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum AccessMode {
+    ReadOnly,
+    ReadWrite,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum MountType {
+    Bind,
+    Volume,
+    Tmpfs,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Mount {
+    pub access_mode: AccessMode,
+    pub mount_type: MountType,
+    pub source: Option<String>,
+    pub target: Option<String>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Host {
+    Addr(IpAddr),
+    HostGateway,
+}
+
+impl ToString for Host {
+    fn to_string(&self) -> String {
+        match self {
+            Host::Addr(ip) => ip.to_string(),
+            Host::HostGateway => "host-gateway".to_string(),
+        }
+    }
+}
+
+type HostPort = u16;
+type ContainerPort = u16;
+pub struct RunnableImage<I: Image> {
+    pub image: I,
+    pub container_name: String,
+    pub overridden_cmd: Option<Vec<String>>,
+    pub network: Option<String>, //todo
+    pub env_vars: BTreeMap<String, String>,
+    pub hosts: BTreeMap<String, Host>,
+    pub mounts: Vec<Mount>, //todo
+    pub ports: BTreeMap<HostPort, ContainerPort>,
+}
+
+impl<I: Image> RunnableImage<I> {
+    pub async fn create_container(
+        &self,
+        client: Arc<Docker>,
+    ) -> anyhow::Result<ContainerId> {
+        all_container_ports_are_exposed(self)?;
+        let config = Config::<String> {
+            image: Some(self.image.image_name().to_string()),
+            host_config: host_config(self),
+            entrypoint: entrypoint(self),
+            cmd: self.overridden_cmd.clone(),
+            env: env(self),
+            ..Default::default()
+        };
+        let options = Some(CreateContainerOptions {
+            name: self.container_name.clone(),
+            platform: None,
+        });
+        match client.create_container(options, config).await {
+            Ok(container) => Ok(container.id),
+            Err(e) => Err(e.into()),
+        }
+    }
+}
+
+/////// helpers ///////
+fn env<I: Image>(runnable_image: &RunnableImage<I>) -> Option<Vec<String>> {
+    match runnable_image.env_vars.is_empty() {
+        true => None,
+        false => Some(
+            runnable_image
+                .env_vars
+                .iter()
+                .map(|(k, v)| format!("{}={}", k, v))
+                .collect::<Vec<_>>(),
+        ),
+    }
+}
+
+fn entrypoint<I: Image>(runnable_image: &RunnableImage<I>) -> Option<Vec<String>> {
+    runnable_image.image.entrypoint().map(|e| {
+        e.split_whitespace()
+            .map(|s| s.to_string())
+            .collect::<Vec<String>>()
+    })
+}
+
+fn host_config<I: Image>(runnable_image: &RunnableImage<I>) -> Option<HostConfig> {
```

Comment:
`HostConfig` doesn't have a None case

```suggestion
fn host_config<I: Image>(runnable_image: &RunnableImage<I>) -> HostConfig> {
```